### ISSUE TITLE
Bundled `gpg` doesn't detect Yubikey

### ISSUE THREAD

**Issue by wojpawlik** [git-for-windows/git#2888]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

   ```
   $ git --version --build-options
   git version 2.29.2.windows.1
   cpu: x86_64
   built from commit: cbf8e7474ba2529d5e7269d0cacab6b4e59579db
   sizeof-long: 4
   sizeof-size_t: 8
   shell-path: /bin/sh
   ```

 - Which version of Windows are you running?

   Windows 10 2004 64-bit

   ```
   $ cmd.exe /c ver

   Microsoft Windows [Version 10.0.19041.572]
   ```

 - What options did you set as part of the installation? Or did you choose the defaults?

   ```
   $ cat /etc/install-options.txt
   Editor Option: VIM
   Custom Editor Path:
   Default Branch Option:
   Path Option: Cmd
   SSH Option: OpenSSH
   Tortoise Option: false
   CURL Option: OpenSSL
   CRLF Option: LFOnly
   Bash Terminal Option: MinTTY
   Git Pull Behavior Option: Rebase
   Use Credential Manager: Core
   Performance Tweaks FSCache: Enabled
   Enable Symlinks: Disabled
   Enable Pseudo Console Support: Enabled
   ```

### Details

 - Which terminal/shell are you running Git from?

   Git Bash

 - What commands did you run to trigger this issue?

   ```sh
   $ gpg --card-status
   ```
 - What did you expect to occur after running these commands?

   ```
   Reader ...........: [REDACTED]
   Application ID ...: [REDACTED]
   Application type .: OpenPGP
   -- snip --
   ```
   Exit code 0

 - What actually happened instead?

   ```
   gpg: selecting card failed: No such device
   gpg: OpenPGP card not available: No such device
   ```
   Exit code 2

https://www.gpg4win.org/ does work correctly, but when I try to use it by renaming the `gpg.exe` bundled with Git for Windows, signing commits works in Git Bash, but not when trying to commit from VSCode GUI.

**Comment by dscho** [comment]
> https://www.gpg4win.org/ does work correctly, but when I try to use it by renaming the `gpg.exe` bundled with Git for Windows, signing commits works in Git Bash, but not when trying to commit from VSCode GUI.

Have you tried configuring [`gpg.program`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-gpgprogram) globally?

**Comment by wojpawlik** [comment]
`$ git config --global gpg.program 'C:/Program Files (x86)/GnuPG/bin/gpg.exe'` and `git config --global gpg.openpgp.program 'C:/Program Files (x86)/GnuPG/bin/gpg.exe'` don't help -- signing works from Git Bash and fails from VSCode with
```
gpg: skipped "Wojciech Pawlik <woj.pawlik@gmail.com>": No secret key
gpg: signing failed: No secret key
error: gpg failed to sign the data
fatal: failed to write commit object
```

**Comment by dscho** [comment]
Can you verify that the intended `gpg.exe` is running (e.g. using Sysinternals' Process Monitor)?

**Comment by wojpawlik** [comment]
Yes, the correct gpg is running.
```
14:39:16.5956997	git.exe	1992	Process Create	C:\Program Files (x86)\GnuPG\bin\gpg.exe	SUCCESS	PID: 5084, Command line: "C:/Program Files (x86)/GnuPG/bin/gpg.exe" -bsau "Wojciech Pawlik <woj.pawlik@gmail.com>"
```

**Comment by dscho** [comment]
But you reported that it works correctly when run in Git Bash? That is very strange. Do you think you can write a little Bash script that runs `gpg.exe` in verbose mode and logs the output into a file, then set your `gpg.program` to that script's path so that we have a chance of learning more about the underlying issue?

**Comment by wojpawlik** [comment]
Windows is such a mess!

I found the issue using Process Explorer. I have Windows and programs on small SSD (`C:`), and everything else on big HDD (`D:`). I modified `$HOME` to point to my main data directory on the HDD, where I wish Git bash to start. Some program didn't respect it, so I changed `$USERPROFILE` to match. Probably as a result of that, some software not utilizing `$APPDATA` and `$LOCALAPPDATA` started putting data in `D:\Wojciech\AppData`. Somehow, gpg4win.org's `gpg-agent` was using `D:` when launched from Git bash (where I imported my key), and `C:` when launched from VSCode. One quick symlink solved the issue.

TL;DR don't mess with environment variables on Windows.

<hr />

Anyway, why doesn't gpg bundled with Git for Windows support smartcards?

**Comment by dscho** [comment]
@wojpawlik I'm glad you figured it out!

> Anyway, why doesn't gpg bundled with Git for Windows support smartcards?

Most likely explanation: nobody worked on that yet ;-) If you want to:

1. install [Git for Windows' SDK](https://gitforwindows.org/#download-sdk),
2. `sdk cd gnupg`,
3. build the package via `sdk build`
4. look around in the log and/or in `src/gnupg-*/` whether you see why smartcards are not supported in that build
5. patch the code to support smartcards (might need a change in the `./configure` call in `/usr/src/MSYS2-packages/gnupg/PKGBUILD`, for example
6. `sdk build`
7. test
8. open a PR?

**Comment by diogocp** [comment]
The Git for Windows version of GnuPG is built with smartcard support. It does not include the CCID driver, but Gpg4win doesn't use it either (it uses PC/SC).

I compared the debug logs from both versions to see where it was failing.

### Git for Windows
> $ /usr/lib/gnupg/scdaemon.exe --server
> enabled debug flags: mpi crypto memory cache memstat hashing ipc cardio reader
> handler for fd -1 started
> DBG: chan_5 -> OK GNU Privacy Guard's Smartcard server ready
> OK GNU Privacy Guard's Smartcard server ready
> serialno
> DBG: chan_5 <- serialno
> DBG: open_pcsc_reader(portstr=(null))
> detected reader 'Yubico Yubikey 4 OTP+U2F+CCID 0'
> reader slot 0: not connected
> DBG: open_pcsc_reader => slot=0
> DBG: enter: apdu_connect: slot=0
> pcsc_control failed: invalid PC/SC error code (0x1)
> pcsc_vendor_specific_init: GET_FEATURE_REQUEST failed: 65547
> reader slot 0: active protocol: T1
> slot 0: ATR=3B F8 13 00 00 81 31 FE 15 59 75 62 69 6B 65 79 34 D4
> DBG: pcsc_get_status_change:  changed present excl inuse
> DBG: leave: apdu_connect => sw=0x0
> DBG: send apdu: c=00 i=A4 p1=00 p2=0C lc=2 le=-1 em=0
> DBG:   PCSC_data: 00 A4 00 0C 02 3F 00
> pcsc_transmit failed: invalid parameter (0x80100004)
> apdu_send_simple(0) failed: invalid value
>[--- REDACTED ---]

### Gpg4win
> $ /c/Program\ Files\ \(x86\)/GnuPG/bin/scdaemon.exe --server
> enabled debug flags: mpi crypto memory cache memstat hashing ipc cardio reader
> handler for fd -1 started
> DBG: chan_0x00000284 -> OK GNU Privacy Guard's Smartcard server ready
> OK GNU Privacy Guard's Smartcard server ready
> serialno
> DBG: chan_0x00000284 <- serialno
> DBG: open_pcsc_reader(portstr=(null))
> detected reader 'Yubico Yubikey 4 OTP+U2F+CCID 0'
> reader slot 0: not connected
> DBG: open_pcsc_reader => slot=0
> DBG: enter: apdu_connect: slot=0
> reader slot 0: active protocol: T1
> slot 0: ATR=3B F8 13 00 00 81 31 FE 15 59 75 62 69 6B 65 79 34 D4
> DBG: pcsc_get_status_change:  changed present excl inuse
> DBG: leave: apdu_connect => sw=0x0
> DBG: send apdu: c=00 i=A4 p1=00 p2=0C lc=2 le=-1 em=0
> DBG:   PCSC_data: 00 A4 00 0C 02 3F 00
> DBG:  response: sw=6D00  datalen=0
> DBG: send apdu: c=00 i=A4 p1=04 p2=00 lc=8 le=-1 em=0
> DBG:   PCSC_data: 00 A4 04 00 08 A0 00 00 05 27 47 11 17
> DBG:  response: sw=9000  datalen=30
>[--- REDACTED ---]

You can see that both versions detect the YubiKey and are able to read the ATR, but the Git for Windows version fails when sending APDUs.

**Comment by diogocp** [comment]
I found this in the GnuPG bug tracker: [scdaemon is not working in cygwin (64bit)](https://web.archive.org/web/20210118084322/https://dev.gnupg.org/T4454).

So I installed the 32-bit version of Git for Windows and guess what, it works fine.

There is a fix but looks like it wasn't backported to the 2.2 branch: gpg/gnupg@1eb93d9229c54baa5f1b7ccf7d105d3692c51a4d.

**Comment by dscho** [comment]
> I found this in the GnuPG bug tracker: [scdaemon is not working in cygwin (64bit)](https://web.archive.org/web/20210118084322/https://dev.gnupg.org/T4454).
> 
> So I installed the 32-bit version of Git for Windows and guess what, it works fine.
> 
> There is a fix but looks like it wasn't backported to the 2.2 branch: gpg/gnupg@1eb93d9229c54baa5f1b7ccf7d105d3692c51a4d.
> 

Wow, good finding!

Could you add that patch to https://github.com/git-for-windows/MSYS2-packages/tree/HEAD/gnupg and open a PR?

You will need to add the patch file, then edit `PKGBUILD` to add it to [the `source` array](https://github.com/git-for-windows/MSYS2-packages/blob/d3dcdfe1de702674bdc6ac5453edd7bb94dd43e4/gnupg/PKGBUILD#L48), the [`sha256sums`](https://github.com/git-for-windows/MSYS2-packages/blob/d3dcdfe1de702674bdc6ac5453edd7bb94dd43e4/gnupg/PKGBUILD#L52), and add a `patch -p1` command to [the `prepare` function](https://github.com/git-for-windows/MSYS2-packages/blob/d3dcdfe1de702674bdc6ac5453edd7bb94dd43e4/gnupg/PKGBUILD#L67).

**Comment by dscho** [comment]
To test the result, I still recommend the steps I outlined here: https://github.com/git-for-windows/git/issues/2888#issuecomment-730301403

**Comment by dscho** [comment]
Hrm. I am trying to verify with my own Yubikey that this works, but I seem to fail, even using 32-bit. @diogocp could you help me diagnose this further, by telling me how exactly I have to hold this thing?

**Comment by diogocp** [comment]
A simple `gpg --card-status` should do the trick. However you may need to enable CCID/OpenPGP mode first using the [YubiKey Manager](https://www.yubico.com/support/download/yubikey-manager/) application (I don't remember if it comes enabled from factory).

**Comment by diogocp** [comment]
Another thing to keep in mind is that if there is already a running scdaemon in the background, gpg will use that one instead of the new one you just built. I had some issues with this so eventually I just uninstalled Git for Windows while testing to be sure that I was running the SDK one.

**Comment by dscho** [comment]
@diogocp could you try again with the just-released v2.33.0?

I _think_ I did everything like you said:

![image](https://user-images.githubusercontent.com/127790/129722677-8befb34c-5d39-489f-a2c3-f1f280ee1ac6.png)

Then in one Git Bash:

```console
$ /usr/lib/gnupg/scdaemon.exe --server --verbose --debug-level 80
scdaemon[451]: enabled debug flags: mpi crypto memory cache memstat ipc cardio reader
scdaemon[451]: handler for fd -1 started
scdaemon[451]: DBG: chan_5 -> OK GNU Privacy Guard's Smartcard server ready
OK GNU Privacy Guard's Smartcard server ready
```

And in another:

```console
$ gpg --card-status --debug-level 80
gpg: enabled debug flags: packet mpi crypto filter iobuf memory cache memstat trust ipc clock lookup extprog
gpg: DBG: [not enabled in the source] start
gpg: DBG: chan_3 <- OK Pleased to meet you, process 482
gpg: DBG: connection to agent established
gpg: DBG: chan_3 -> RESET
gpg: DBG: chan_3 <- OK
gpg: DBG: chan_3 -> OPTION ttyname=/dev/pty0
gpg: DBG: chan_3 <- OK
gpg: DBG: chan_3 -> OPTION ttytype=xterm
gpg: DBG: chan_3 <- OK
gpg: DBG: chan_3 -> OPTION display=needs-to-be-defined
gpg: DBG: chan_3 <- OK
gpg: DBG: chan_3 -> OPTION lc-ctype=en_US.UTF-8
gpg: DBG: chan_3 <- OK
gpg: DBG: chan_3 -> OPTION lc-messages=C
gpg: DBG: chan_3 <- OK
gpg: DBG: chan_3 -> GETINFO version
gpg: DBG: chan_3 <- D 2.2.29-unknown
gpg: DBG: chan_3 <- OK
gpg: DBG: chan_3 -> OPTION allow-pinentry-notify
gpg: DBG: chan_3 <- OK
gpg: DBG: chan_3 -> OPTION agent-awareness=2.1.0
gpg: DBG: chan_3 <- OK
gpg: DBG: chan_3 -> SCD GETINFO version
gpg: DBG: chan_3 <- D 2.2.29-unknown
gpg: DBG: chan_3 <- OK
gpg: DBG: chan_3 -> SCD SERIALNO
gpg: DBG: chan_3 <- ERR 100696144 No such device <SCD>
gpg: selecting card failed: No such device
gpg: OpenPGP card not available: No such device
gpg: DBG: [not enabled in the source] stop
gpg: keydb: handles=0 locks=0 parse=0 get=0
gpg:        build=0 update=0 insert=0 delete=0
gpg:        reset=0 found=0 not=0 cache=0 not=0
gpg: kid_not_found_cache: count=0 peak=0 flushes=0
gpg: sig_cache: total=0 cached=0 good=0 bad=0
gpg: random usage: poolsize=600 mixed=0 polls=0/0 added=0/0
              outmix=0 getlvl1=0/0 getlvl2=0/0
gpg: rndjent stat: collector=0x0000000000000000 calls=0 bytes=0
gpg: secmem usage: 0/65536 bytes in 0 blocks
```

It is a bit strange that I do not see any further output in the `scdaemon` window, either.

Any ideas?

**Comment by diogocp** [comment]
Just tested v2.33.0 and it's working perfectly.

Try the following:
1. Terminate existing scdaemon: `echo scd killscd | gpg-connect-agent --decode`
2. Start scdaemon with debug logs: `/usr/lib/gnupg/scdaemon.exe --server --debug-level guru`
3. Type `serialno` (enter). This will show you logs similar to the ones in [my initial comment](#issuecomment-899064872).

>serialno
>scdaemon[1475]: DBG: chan_5 <- serialno
>scdaemon[1475]: DBG: open_pcsc_reader(portstr=(null))
>scdaemon[1475]: detected reader 'Yubico Yubikey 4 OTP+U2F+CCID 0'
>scdaemon[1475]: reader slot 0: not connected
>scdaemon[1475]: DBG: open_pcsc_reader => slot=0
>scdaemon[1475]: DBG: enter: apdu_connect: slot=0
>scdaemon[1475]: pcsc_control failed: invalid PC/SC error code (0x1)
>scdaemon[1475]: pcsc_vendor_specific_init: GET_FEATURE_REQUEST failed: 65547
>scdaemon[1475]: reader slot 0: active protocol: T1
>scdaemon[1475]: slot 0: ATR=3B F8 13 00 00 81 31 FE 15 59 75 62 69 6B 65 79 34 D4
>scdaemon[1475]: DBG: pcsc_get_status_change:  changed present excl inuse
>scdaemon[1475]: DBG: leave: apdu_connect => sw=0x0
>scdaemon[1475]: DBG: send apdu: c=00 i=A4 p1=00 p2=0C lc=2 le=-1 em=0
>scdaemon[1475]: DBG:   PCSC_data: 00 A4 00 0C 02 3F 00
>scdaemon[1475]: DBG:  response: sw=6D00  datalen=0
>scdaemon[1475]: DBG: send apdu: c=00 i=A4 p1=04 p2=00 lc=8 le=-1 em=0
>scdaemon[1475]: DBG:   PCSC_data: 00 A4 04 00 08 A0 00 00 05 27 47 11 17
>scdaemon[1475]: DBG:  response: sw=9000  datalen=30
>[redacted]

The key things to look for are these lines:
>scdaemon[1475]: detected reader 'Yubico Yubikey 4 OTP+U2F+CCID 0'

(reader detected)

>scdaemon[1475]: slot 0: ATR=3B F8 13 00 00 81 31 FE 15 59 75 62 69 6B 65 79 34 D4

(card reset and ATR received)

>scdaemon[1475]: DBG:  response: sw=9000  datalen=30

(card responded OK to APDU)

**Comment by dscho** [comment]
I get this:

```console
$ /usr/lib/gnupg/scdaemon.exe --server --debug-level guru
scdaemon[500]: enabled debug flags: mpi crypto memory cache memstat hashing ipc cardio reader
scdaemon[500]: handler for fd -1 started
scdaemon[500]: DBG: chan_5 -> OK GNU Privacy Guard's Smartcard server ready
OK GNU Privacy Guard's Smartcard server ready
serialno
scdaemon[500]: DBG: chan_5 <- serialno
scdaemon[500]: DBG: open_pcsc_reader(portstr=(null))
scdaemon[500]: detected reader 'Windows Hello for Business 1'
scdaemon[500]: detected reader 'Yubico YubiKey FIDO+CCID 0'
scdaemon[500]: reader slot 0: not connected
scdaemon[500]: DBG: open_pcsc_reader => slot=0
scdaemon[500]: DBG: enter: apdu_connect: slot=0
scdaemon[500]: pcsc_control failed: unknown PC/SC error code (0x80100022)
scdaemon[500]: pcsc_vendor_specific_init: GET_FEATURE_REQUEST failed: 65547
scdaemon[500]: reader slot 0: active protocol: T1
scdaemon[500]: slot 0: ATR=3B 8D 01 80 FB A0 00 00 03 97 42 54 46 59 04 01 CF
scdaemon[500]: DBG: pcsc_get_status_change:  changed present excl inuse
scdaemon[500]: DBG: leave: apdu_connect => sw=0x0
scdaemon[500]: DBG: send apdu: c=00 i=A4 p1=00 p2=0C lc=2 le=-1 em=0
scdaemon[500]: DBG:   PCSC_data: 00 A4 00 0C 02 3F 00
scdaemon[500]: DBG:  response: sw=6A82  datalen=0
scdaemon[500]: DBG: enter: apdu_get_atr: slot=0
scdaemon[500]: DBG: leave: apdu_get_atr => atrlen=17
scdaemon[500]: DBG: send apdu: c=00 i=A4 p1=04 p2=00 lc=6 le=-1 em=0
scdaemon[500]: DBG:   PCSC_data: 00 A4 04 00 06 D2 76 00 01 24 01
scdaemon[500]: DBG:  response: sw=6A82  datalen=0
scdaemon[500]: DBG: send apdu: c=00 i=A4 p1=04 p2=0C lc=7 le=-1 em=0
scdaemon[500]: DBG:   PCSC_data: 00 A4 04 0C 07 D2 76 00 00 03 01 02
scdaemon[500]: DBG:  response: sw=6A82  datalen=0
scdaemon[500]: DBG: send apdu: c=00 i=A4 p1=04 p2=00 lc=12 le=256 em=0
scdaemon[500]: DBG:   PCSC_data: 00 A4 04 00 0C A0 00 00 00 63 50 4B 43 53 2D 31 35 00
scdaemon[500]: DBG:  response: sw=6A82  datalen=0
scdaemon[500]: DBG: send apdu: c=00 i=A4 p1=08 p2=0C lc=2 le=-1 em=0
scdaemon[500]: DBG:   PCSC_data: 00 A4 08 0C 02 2F 00
scdaemon[500]: DBG:  response: sw=6A86  datalen=0
scdaemon[500]: DBG: send apdu: c=00 i=A4 p1=01 p2=0C lc=2 le=-1 em=0
scdaemon[500]: DBG:   PCSC_data: 00 A4 01 0C 02 50 15
scdaemon[500]: DBG:  response: sw=6A86  datalen=0
scdaemon[500]: DBG: send apdu: c=00 i=A4 p1=04 p2=0C lc=9 le=-1 em=0
scdaemon[500]: DBG:   PCSC_data: 00 A4 04 0C 09 D2 76 00 00 25 45 50 02 00
scdaemon[500]: DBG:  response: sw=6A82  datalen=0
scdaemon[500]: DBG: send apdu: c=00 i=A4 p1=04 p2=0C lc=6 le=-1 em=0
scdaemon[500]: DBG:   PCSC_data: 00 A4 04 0C 06 D2 76 00 00 66 01
scdaemon[500]: DBG:  response: sw=6A82  datalen=0
scdaemon[500]: DBG: send apdu: c=00 i=A4 p1=04 p2=0C lc=11 le=-1 em=0
scdaemon[500]: DBG:   PCSC_data: 00 A4 04 0C 0B E8 2B 06 01 04 01 81 C3 1F 02 01
scdaemon[500]: DBG:  response: sw=6A82  datalen=0
scdaemon[500]: no supported card application found: No such file or directory
scdaemon[500]: DBG: enter: apdu_close_reader: slot=0
scdaemon[500]: DBG: enter: apdu_disconnect: slot=0
scdaemon[500]: DBG: leave: apdu_disconnect => sw=0x0
scdaemon[500]: DBG: leave: apdu_close_reader => 0x0 (close_reader)
scdaemon[500]: DBG: chan_5 -> ERR 100696144 No such device <SCD>
ERR 100696144 No such device <SCD>
```

**Comment by dscho** [comment]
At first, I thought that "no supported card application found" might mean that I excluded too many files from being bundled in Git for Windows, so I tried with an up-to-date SDK, but I get the same:

```console
$ /usr/lib/gnupg/scdaemon.exe --server --debug-level guru
scdaemon[1084]: enabled debug flags: mpi crypto memory cache memstat hashing ipc cardio reader
scdaemon[1084]: handler for fd -1 started
scdaemon[1084]: DBG: chan_5 -> OK GNU Privacy Guard's Smartcard server ready
OK GNU Privacy Guard's Smartcard server ready
serialno
scdaemon[1084]: DBG: chan_5 <- serialno
scdaemon[1084]: DBG: open_pcsc_reader(portstr=(null))
scdaemon[1084]: detected reader 'Windows Hello for Business 1'
scdaemon[1084]: detected reader 'Yubico YubiKey FIDO+CCID 0'
scdaemon[1084]: reader slot 0: not connected
scdaemon[1084]: DBG: open_pcsc_reader => slot=0
scdaemon[1084]: DBG: enter: apdu_connect: slot=0
scdaemon[1084]: pcsc_control failed: unknown PC/SC error code (0x80100022)
scdaemon[1084]: pcsc_vendor_specific_init: GET_FEATURE_REQUEST failed: 65547
scdaemon[1084]: reader slot 0: active protocol: T1
scdaemon[1084]: slot 0: ATR=3B 8D 01 80 FB A0 00 00 03 97 42 54 46 59 04 01 CF
scdaemon[1084]: DBG: pcsc_get_status_change:  changed present excl inuse
scdaemon[1084]: DBG: leave: apdu_connect => sw=0x0
scdaemon[1084]: DBG: send apdu: c=00 i=A4 p1=00 p2=0C lc=2 le=-1 em=0
scdaemon[1084]: DBG:   PCSC_data: 00 A4 00 0C 02 3F 00
scdaemon[1084]: DBG:  response: sw=6A82  datalen=0
scdaemon[1084]: DBG: enter: apdu_get_atr: slot=0
scdaemon[1084]: DBG: leave: apdu_get_atr => atrlen=17
scdaemon[1084]: DBG: send apdu: c=00 i=A4 p1=04 p2=00 lc=6 le=-1 em=0
scdaemon[1084]: DBG:   PCSC_data: 00 A4 04 00 06 D2 76 00 01 24 01
scdaemon[1084]: DBG:  response: sw=6A82  datalen=0
scdaemon[1084]: DBG: send apdu: c=00 i=A4 p1=04 p2=0C lc=7 le=-1 em=0
scdaemon[1084]: DBG:   PCSC_data: 00 A4 04 0C 07 D2 76 00 00 03 01 02
scdaemon[1084]: DBG:  response: sw=6A82  datalen=0
scdaemon[1084]: DBG: send apdu: c=00 i=A4 p1=04 p2=00 lc=12 le=256 em=0
scdaemon[1084]: DBG:   PCSC_data: 00 A4 04 00 0C A0 00 00 00 63 50 4B 43 53 2D 31 35 00
scdaemon[1084]: DBG:  response: sw=6A82  datalen=0
scdaemon[1084]: DBG: send apdu: c=00 i=A4 p1=08 p2=0C lc=2 le=-1 em=0
scdaemon[1084]: DBG:   PCSC_data: 00 A4 08 0C 02 2F 00
scdaemon[1084]: DBG:  response: sw=6A86  datalen=0
scdaemon[1084]: DBG: send apdu: c=00 i=A4 p1=01 p2=0C lc=2 le=-1 em=0
scdaemon[1084]: DBG:   PCSC_data: 00 A4 01 0C 02 50 15
scdaemon[1084]: DBG:  response: sw=6A86  datalen=0
scdaemon[1084]: DBG: send apdu: c=00 i=A4 p1=04 p2=0C lc=9 le=-1 em=0
scdaemon[1084]: DBG:   PCSC_data: 00 A4 04 0C 09 D2 76 00 00 25 45 50 02 00
scdaemon[1084]: DBG:  response: sw=6A82  datalen=0
scdaemon[1084]: DBG: send apdu: c=00 i=A4 p1=04 p2=0C lc=6 le=-1 em=0
scdaemon[1084]: DBG:   PCSC_data: 00 A4 04 0C 06 D2 76 00 00 66 01
scdaemon[1084]: DBG:  response: sw=6A82  datalen=0
scdaemon[1084]: DBG: send apdu: c=00 i=A4 p1=04 p2=0C lc=11 le=-1 em=0
scdaemon[1084]: DBG:   PCSC_data: 00 A4 04 0C 0B E8 2B 06 01 04 01 81 C3 1F 02 01
scdaemon[1084]: DBG:  response: sw=6A82  datalen=0
scdaemon[1084]: no supported card application found: No such file or directory
scdaemon[1084]: DBG: enter: apdu_close_reader: slot=0
scdaemon[1084]: DBG: enter: apdu_disconnect: slot=0
scdaemon[1084]: DBG: leave: apdu_disconnect => sw=0x0
scdaemon[1084]: DBG: leave: apdu_close_reader => 0x0 (close_reader)
scdaemon[1084]: DBG: chan_5 -> ERR 100696144 No such device <SCD>
ERR 100696144 No such device <SCD>
```

**Comment by diogocp** [comment]
>scdaemon[1084]: detected reader 'Windows Hello for Business 1'

That rings a bell ... I did not have the problem myself, but I stumbled upon this while investigating my scdaemon issues: whilehttps://superuser.com/questions/1343569/scdaemon-wont-detect-yubikey-on-windows-10

You should be able to fix it by setting `reader-port Yubico Yubi` in `~/.gnupg/scdaemon.conf`.
See also this: https://support.yubico.com/hc/en-us/articles/360013714479-Troubleshooting-Issues-with-GPG

**Comment by dscho** [comment]
Indeed, with `/usr/lib/gnupg/scdaemon.exe --server --reader-port "Yubico Yubi" --debug-level guru` I come a lot further ;-)

**Comment by dscho** [comment]
> You should be able to fix it by setting `reader-port Yubico Yubi` in `~/.gnupg/scdaemon.conf`.

And with this, `gpg --card-status` works, too! Yay!
